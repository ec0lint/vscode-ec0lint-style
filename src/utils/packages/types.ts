// eslint-disable-next-line node/no-unpublished-import
import type Ec0lintStyle from 'ec0lint-style';

/**
 * Package manager identifiers.
 */
export type PackageManager = 'npm' | 'yarn' | 'pnpm';

/**
 * Options for resolving the Stylelint package.
 */
export type ResolverOptions = {
	packageManager?: PackageManager;
	stylelintPath?: string;
};

/**
 * Stylelint package resolution result.
 */
export type StylelintResolutionResult = {
	stylelint: Ec0lintStyle.PublicApi;
	resolvedPath: string;
};

/**
 * A tracer function that can be used to log messages.
 */
export type TracerFn = (message: string, verbose?: string) => void;
