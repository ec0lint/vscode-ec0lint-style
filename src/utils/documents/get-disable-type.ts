import { Position } from 'vscode-languageserver-types';
import type { TextDocument } from 'vscode-languageserver-textdocument';

/**
 * Disable directive comment types.
 */
export type DisableType =
	| 'ec0lint-style-disable'
	| 'ec0lint-style-disable-line'
	| 'ec0lint-style-disable-next-line';

/**
 * If the given position is inside a `ec0lint-style-disable` after the comment'
 * type, returns the disable comment's type. Otherwise, returns `undefined`.
 *
 * @example
 * ```js
 * const document = TextDocument.create(
 *   'file:///path/to/file.css',
 *   'css',
 *   1,
 *   '/* ec0lint-style-disable-line indentation *\/'
 *   //                         ^ Position is here
 * );
 * const position = Position.create(0, 26);
 *
 * getDisableType(document, position);
 * // => 'ec0lint-style-disable-line'
 * ```
 */
export function getDisableType(
	document: TextDocument,
	position: Position,
): DisableType | undefined {
	const lineStartOffset = document.offsetAt(Position.create(position.line, 0));
	const lineEndOffset = document.offsetAt(Position.create(position.line + 1, 0));
	const line = document.getText().slice(lineStartOffset, lineEndOffset);

	const before = line.slice(0, position.character);
	const after = line.slice(position.character);

	const disableKind = before
		.match(/\/\*\s*(ec0lint-style-disable(?:(?:-next)?-line)?)\s[a-z\-/\s,]*$/i)?.[1]
		?.toLowerCase();

	return disableKind && /^[a-z\-/\s,]*\*\//i.test(after) ? (disableKind as DisableType) : undefined;
}
