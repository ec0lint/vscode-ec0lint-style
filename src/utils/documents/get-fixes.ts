import { createTextEdits } from './create-text-edits';
// eslint-disable-next-line node/no-unpublished-import
import type Ec0lintStyle from 'ec0lint-style';
import type { TextDocument, TextEdit } from 'vscode-languageserver-textdocument';
import type { Ec0lintRunner, RunnerOptions } from '../ec0lint-style';

/**
 * Runs Ec0lint and returns fix text edits for the given document.
 * @param runner The Ec0lint runner.
 * @param document The document to get fixes for.
 * @param linterOptions Linter options to use.
 * @param runnerOptions The runner options.
 */
export async function getFixes(
	runner: Ec0lintRunner,
	document: TextDocument,
	linterOptions: Ec0lintStyle.LinterOptions = {},
	runnerOptions: RunnerOptions = {},
): Promise<TextEdit[]> {
	const result = await runner.lintDocument(
		document,
		{ ...linterOptions, fix: true },
		runnerOptions,
	);

	return typeof result.output === 'string' ? createTextEdits(document, result.output) : [];
}
