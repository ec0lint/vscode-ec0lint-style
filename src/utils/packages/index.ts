export * from './find-package-root';
export * from './global-path-resolver';
export * from './ec0lint-style-resolver';
export * from './types';
