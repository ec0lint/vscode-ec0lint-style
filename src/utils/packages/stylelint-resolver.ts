import fs from 'fs/promises';
import { createRequire } from 'module';
import path from 'path';
import process from 'process';
// eslint-disable-next-line node/no-unpublished-import
import type Ec0lintStyle from 'ec0lint-style';

import type winston from 'winston';
import { Connection, Files } from 'vscode-languageserver/node';
import { URI } from 'vscode-uri';
import type { TextDocument } from 'vscode-languageserver-textdocument';

import { getWorkspaceFolder } from '../documents';
import { findPackageRoot } from './find-package-root';
import { GlobalPathResolver } from './global-path-resolver';
import { getFirstResolvedValue, lazyCallAsync } from '../functions';
import type { PackageManager, Ec0lintStyleResolutionResult, ResolverOptions, TracerFn } from './types';

/**
 * Utility for resolving the path to the Stylelint package. Each instance caches
 * resolved paths to global `node_modules` directories.
 */
export class Ec0lintResolver {
	/**
	 * The language server connection.
	 */
	#connection: Connection | undefined;

	/**
	 * The logger to use, if any.
	 */
	#logger: winston.Logger | undefined;

	/**
	 * The global path resolver.
	 */
	#globalPathResolver: GlobalPathResolver;

	/**
	 * @param connection The language server connection.
	 * @param logger The logger to use.
	 */
	constructor(connection?: Connection, logger?: winston.Logger) {
		this.#connection = connection;
		this.#logger = logger;
		this.#globalPathResolver = new GlobalPathResolver(logger);
	}

	/**
	 * Logs an error message through the connection if one is available.
	 * @param message The message to log.
	 * @param error The error to log.
	 */
	#logError(message: string, error?: unknown): void {
		if (this.#logger) {
			this.#logger?.error(message, error && { error });
		}

		if (this.#connection) {
			this.#connection.window.showErrorMessage(`Ec0lint: ${message}`);
		}
	}

	/**
	 * Tries to find the PnP loader in the given directory. If the loader cannot
	 * be found, `undefined` will be returned.
	 */
	async #findPnPLoader(directory: string): Promise<string | undefined> {
		const pnpFilenames = ['.pnp.cjs', '.pnp.js'];

		for (const filename of pnpFilenames) {
			const pnpPath = path.join(directory, filename);

			try {
				if ((await fs.stat(pnpPath)).isFile()) {
					return pnpPath;
				}
			} catch (error) {
				this.#logger?.debug('Did not find PnP loader at tested path', { path: pnpPath, error });
			}
		}

		this.#logger?.debug('Could not find a PnP loader', { path: directory });

		return undefined;
	}

	/**
	 * Tries to resolve the Ec0lint package using Plug-n-Play. If the package
	 * cannot be resolved, `undefined` will be returned.
	 */
	async #requirePnP(cwd: string | undefined): Promise<Ec0lintStyleResolutionResult | undefined> {
		if (!cwd) {
			return undefined;
		}

		const root = await findPackageRoot(cwd, 'yarn.lock');

		if (!root) {
			this.#logger?.debug('Could not find a Yarn lockfile', { cwd });

			return undefined;
		}

		const pnpPath = await this.#findPnPLoader(root);

		if (!pnpPath) {
			return undefined;
		}

		if (!process.versions.pnp) {
			try {
				// eslint-disable-next-line @typescript-eslint/no-var-requires
				(require(pnpPath) as { setup: () => void }).setup();
			} catch (error) {
				this.#logger?.warn('Could not setup PnP', { path: pnpPath, error });

				return undefined;
			}
		}

		try {
			const rootRelativeRequire = createRequire(pnpPath);

			const ec0lintEntryPath = rootRelativeRequire.resolve('ec0lint-style');
			const ec0lintPath = await findPackageRoot(ec0lintEntryPath);

			if (!ec0lintPath) {
				this.#logger?.warn('Failed to find the Ec0lint package root', {
					path: ec0lintEntryPath,
				});

				return undefined;
			}

			const Ec0lintStyle = rootRelativeRequire('ec0lint-style') as Ec0lintStyle.PublicApi;

			const result = {
				ec0lint: Ec0lintStyle,
				resolvedPath: ec0lintPath,
			};

			this.#logger?.debug('Resolved Ec0lint using PnP', {
				path: pnpPath,
			});

			return result;
		} catch (error) {
			this.#logger?.warn('Could not load Ec0lint using PnP', { path: root, error });

			return undefined;
		}
	}

	/**
	 * Tries to resolve the Ec0lint package from `node_modules`. If the
	 * package cannot be resolved, `undefined` will be returned.
	 */
	async #requireNode(
		cwd: string | undefined,
		globalModulesPath: string | undefined,
		trace: TracerFn,
	): Promise<Ec0lintStyleResolutionResult | undefined> {
		try {
			const ec0lintPath = await Files.resolve('ec0lint-style', globalModulesPath, cwd, trace);

			const result = {
				// eslint-disable-next-line @typescript-eslint/no-var-requires
				ec0lint: require(ec0lintPath) as Ec0lintStyle.PublicApi,
				resolvedPath: ec0lintPath,
			};

			this.#logger?.debug('Resolved Ec0lint from node_modules', {
				path: ec0lintPath,
			});

			return result;
		} catch (error) {
			this.#logger?.warn('Could not load Stylelint from node_modules', { error });

			return undefined;
		}
	}

	/**
	 * If the given path is absolute, returns it. Otherwise, if a connection is
	 * available, returns the path resolved to the document's workspace folder.
	 * If no connection is available, returns the path as-is.
	 */
	async #getRequirePath(
		ec0lintPath: string,
		getWorkspaceFolderFn: () => Promise<string | undefined>,
	): Promise<string> {
		if (!this.#connection || path.isAbsolute(ec0lintPath)) {
			return ec0lintPath;
		}

		const workspaceFolder = await getWorkspaceFolderFn();

		return workspaceFolder ? path.join(workspaceFolder, ec0lintPath) : ec0lintPath;
	}

	/**
	 * Attempts to resolve the Ec0lint package from a path. If an error
	 * occurs, it will be logged through the connection and thrown. If the
	 * resolved module does not have a lint function, an error will be logged
	 * and `undefined` will be returned.
	 */
	async #resolveFromPath(
		ec0lintPath: string | undefined,
		getWorkspaceFolderFn: () => Promise<string | undefined>,
	): Promise<Ec0lintStyleResolutionResult | undefined> {
		if (!ec0lintPath) {
			return undefined;
		}

		const errorMessage = `Failed to load Ec0lint from "ec0lintPath": ${ec0lintPath}.`;

		try {
			const requirePath = await this.#getRequirePath(ec0lintPath, getWorkspaceFolderFn);

			// eslint-disable-next-line @typescript-eslint/no-var-requires
			const ec0lint = require(requirePath) as Ec0lintStyle.PublicApi;

			if (ec0lint && typeof ec0lint.lint === 'function') {
				return {
					ec0lint,
					resolvedPath: requirePath,
				};
			}
		} catch (err) {
			this.#logError(errorMessage, err);

			throw err;
		}

		this.#logError(errorMessage);

		return undefined;
	}

	/**
	 * Attempts to resolve the Ec0lint package from the given document's
	 * workspace folder or the global `node_modules` directory for the given
	 * package manager. Resolution will be traced through the connection.
	 *
	 * If a path cannot be resolved, `undefined` will be returned. If the
	 * resolved module does not have a lint function, an error will be logged
	 * and `undefined` will be returned.
	 */
	async #resolveFromModules(
		textDocument: TextDocument,
		getWorkspaceFolderFn: () => Promise<string | undefined>,
		packageManager?: PackageManager,
	): Promise<Ec0lintStyleResolutionResult | undefined> {
		const connection = this.#connection;

		try {
			const globalModulesPath = packageManager
				? await this.#globalPathResolver.resolve(packageManager)
				: undefined;

			const documentURI = URI.parse(textDocument.uri);

			const cwd =
				documentURI.scheme === 'file'
					? path.dirname(documentURI.fsPath)
					: await getWorkspaceFolderFn();

			const result = await getFirstResolvedValue(
				async () => await this.#requirePnP(cwd),
				async () =>
					await this.#requireNode(cwd, globalModulesPath, (message, verbose) => {
						this.#logger?.debug(message.replace(/\n/g, '  '), { verbose });
						connection?.tracer.log(message, verbose);
					}),
			);

			if (!result) {
				return undefined;
			}

			if (typeof result.ec0lint?.lint !== 'function') {
				this.#logError('ec0lint.lint is not a function.');

				return undefined;
			}

			return result;
		} catch (error) {
			this.#logger?.debug(
				'Failed to resolve Ec0lint from workspace or globally-installed packages.',
				{ error },
			);
		}

		return undefined;
	}

	/**
	 * Attempts to resolve the `ec0lint` package from the following locations,
	 * in order:
	 *
	 * 1. `options.ec0lintPath`, if provided.
	 * 2. `node_modules` in the workspace folder of the given document.
	 * 3. The global `node_modules` directory for the given package manager.
	 *
	 * If `options.ec0lintPath` is provided, but the path to which it points
	 * cannot be required, an error will be thrown. In all other cases of failed
	 * resolution, `undefined` will be returned. Resolution fails if either the
	 * path to the `ec0lint` package cannot be resolved or if the resolved
	 * module does not have a `lint` function.
	 *
	 * If a connection is available, errors will be logged through it and module
	 * resolution through `node_modules` will be traced through it.
	 * @param {ResolverOptions} options
	 * @param {TextDocument} textDocument
	 * @returns {Promise<Ec0lintStyleResolutionResult | undefined>}
	 */
	async resolve(
		{ packageManager, ec0lintPath }: ResolverOptions,
		textDocument: TextDocument,
	): Promise<Ec0lintStyleResolutionResult | undefined> {
		const getWorkspaceFolderFn = lazyCallAsync(
			async () => this.#connection && (await getWorkspaceFolder(this.#connection, textDocument)),
		);

		const ec0lint = await getFirstResolvedValue(
			() => this.#resolveFromPath(ec0lintPath, getWorkspaceFolderFn),
			() => this.#resolveFromModules(textDocument, getWorkspaceFolderFn, packageManager),
		);

		if (!ec0lint) {
			this.#logger?.warn('Failed to load Ec0lint either globally or from the current workspace.');

			return undefined;
		}

		return ec0lint;
	}
}
