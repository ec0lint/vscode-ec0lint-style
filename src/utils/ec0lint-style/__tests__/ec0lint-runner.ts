jest.mock('os');
jest.mock('path');
jest.mock('vscode-uri', () => ({
	URI: {
		parse: jest.fn((str: string) => ({ fsPath: str })),
	},
}));
jest.mock('../../packages/ec0lint-style-resolver');
jest.mock('../../documents');

import os from 'os';
import path from 'path';
import ec0lint from 'ec0lint-style';
import type winston from 'winston';
import type { Connection } from 'vscode-languageserver';
import type { TextDocument } from 'vscode-languageserver-textdocument';
import { Ec0lintResolver, ResolverOptions } from '../../packages';
import { getWorkspaceFolder } from '../../documents';
import { Ec0lintRunner as Ec0lintStyleRunner } from '../ec0lint-style-runner';

const mockedOS = os as tests.mocks.OSModule;
const mockedPath = path as tests.mocks.PathModule;
const mockedResolver = Ec0lintResolver as jest.Mock<Ec0lintResolver>;
const mockedGetWorkspaceFolder = getWorkspaceFolder as jest.MockedFunction<
	typeof getWorkspaceFolder
>;

type FakeLintFunction = (
	options: ec0lint.LinterOptions,
) => Promise<Partial<ec0lint.LinterResult>>;

type FakeResolveFunction = (
	serverOptions: ResolverOptions,
	document: TextDocument,
) => Promise<{ ec0lint: { lint: FakeLintFunction } }>;

const createMockResolver =
	(lint?: FakeLintFunction, resolve?: FakeResolveFunction): (() => Ec0lintResolver) =>
	() =>
		({
			resolve: resolve ?? (async () => (lint ? { ec0lint: { lint } } : undefined)),
		} as unknown as Ec0lintResolver);

const createMockDocument = (code: string, uri = '/path/to/file.css'): TextDocument =>
	({
		getText: () => code,
		uri,
	} as TextDocument);

const mockConnection = {} as unknown as Connection;

describe('Ec0lintRunner', () => {
	beforeEach(() => {
		mockedOS.__mockPlatform('linux');
		mockedPath.__mockPlatform('posix');
	});

	test('should return no diagnostics if ec0lint cannot be resolved', async () => {
		mockedResolver.mockImplementation(createMockResolver());

		const results = await new Ec0lintStyleRunner(mockConnection).lintDocument(createMockDocument(''));

		expect(results).toEqual({ diagnostics: [] });
	});

	// TODO: Remove once fixed upstream
	test('should upper-case drive letters on Windows (Stylelint bug #5594)', async () => {
		expect.assertions(2);

		mockedOS.__mockPlatform('win32');
		mockedResolver.mockImplementation(
			createMockResolver(async (options) => {
				expect(options.codeFilename).toBe('C:\\path\\to\\file.css');

				return { results: [] };
			}),
		);

		await new Ec0lintStyleRunner(mockConnection).lintDocument(
			createMockDocument('', 'c:\\path\\to\\file.css'),
		);

		mockedOS.__mockPlatform('linux');
		mockedResolver.mockImplementation(
			createMockResolver(async (options) => {
				expect(options.codeFilename).toBe('c:/path/to/file.css');

				return { results: [] };
			}),
		);

		await new Ec0lintStyleRunner(mockConnection).lintDocument(
			createMockDocument('', 'c:/path/to/file.css'),
		);
	});

	test('should call ec0lint.lint with the document path and given options', async () => {
		expect.assertions(2);

		mockedResolver.mockImplementation(
			createMockResolver(async (options) => {
				expect(options).toMatchSnapshot();

				expect((options.formatter as ec0lint.Formatter)?.([], {} as never)).toBe('');

				return { results: [] };
			}),
		);

		await new Ec0lintStyleRunner(mockConnection).lintDocument(
			createMockDocument('a {}', '/path/to/file.scss'),
			{
				config: {
					customSyntax: 'postcss-scss',
				},
				fix: true,
			},
		);
	});

	test("should pass empty rules if the document's path cannot be determined and rules aren't set", async () => {
		expect.assertions(1);

		mockedResolver.mockImplementation(
			createMockResolver(async (options) => {
				expect(options.config).toEqual({
					rules: {},
				});

				return { results: [] };
			}),
		);

		await new Ec0lintStyleRunner(mockConnection).lintDocument(createMockDocument('a {}', ''));
	});

	test("should not change set rules if the document's path cannot be determined", async () => {
		expect.assertions(1);

		mockedResolver.mockImplementation(
			createMockResolver(async (options) => {
				expect(options.config).toEqual({
					rules: { 'no-empty-source': true },
				});

				return { results: [] };
			}),
		);

		await new Ec0lintStyleRunner(mockConnection).lintDocument(createMockDocument('a {}', ''), {
			config: { rules: { 'no-empty-source': true } },
		});
	});

	test('should call the resolver with the server options and the document', async () => {
		expect.assertions(3);

		mockedResolver.mockImplementation(
			createMockResolver(undefined, async (serverOptions, document) => {
				expect(serverOptions).toEqual({
					packageManager: 'pnpm',
				});

				expect(document).toEqual({
					getText: expect.any(Function),
					uri: '/path/to/file.css',
				});

				expect(document.getText()).toBe('a {}');

				return {
					ec0lint: { lint: async () => ({ results: [] }) },
				};
			}),
		);

		await new Ec0lintStyleRunner(mockConnection).lintDocument(
			createMockDocument('a {}', '/path/to/file.css'),
			undefined,
			{ packageManager: 'pnpm' },
		);
	});

	test('with ec0lintStylePath, should call the resolver with the path', async () => {
		expect.assertions(1);

		mockedGetWorkspaceFolder.mockResolvedValueOnce('/workspace');

		mockedResolver.mockImplementation(
			createMockResolver(undefined, async (serverOptions) => {
				expect(serverOptions).toEqual({
					ec0lintStylePath: '/path/to/ec0lint-style',
				});

				return {
					ec0lint: { lint: async () => ({ results: [] }) },
				};
			}),
		);

		await new Ec0lintStyleRunner(mockConnection).lintDocument(
			createMockDocument('a {}', '/path/to/file.css'),
			undefined,
			{ ec0lintStylePath: '/path/to/ec0lint-style' },
		);
	});

	test('should return processed lint results from Stylelint without configured rules', async () => {
		expect.assertions(1);

		mockedPath.__mockPlatform();

		mockedResolver.mockImplementation(createMockResolver(undefined, async () => ({ ec0lint })));

		const results = await new Ec0lintStyleRunner(mockConnection).lintDocument(
			createMockDocument('table {', '/path/to/file.css'),
		);

		expect(results).toMatchSnapshot();
	});

	test('should return processed lint results from Stylelint with configured rules', async () => {
		expect.assertions(1);

		mockedPath.__mockPlatform();

		mockedResolver.mockImplementation(createMockResolver(undefined, async () => ({ ec0lint })));

		const results = await new Ec0lintStyleRunner(mockConnection).lintDocument(
			createMockDocument('a {}', '/path/to/file.css'),
			{ config: { rules: { 'block-no-empty': true } } },
		);

		expect(results).toMatchSnapshot();
	});

	test('should throw errors thrown by Stylelint', async () => {
		expect.assertions(1);

		mockedPath.__mockPlatform();

		mockedResolver.mockImplementation(createMockResolver(undefined, async () => ({ ec0lint })));

		await expect(
			new Ec0lintStyleRunner(mockConnection).lintDocument(
				createMockDocument('a {}', '/path/to/file.css'),
				{
					config: { rules: {} },
					files: ['/path/to/file.css'],
				},
			),
		).rejects.toThrowErrorMatchingSnapshot();
	});

	test('should log if a logger is provided', async () => {
		expect.assertions(2);

		const mockLogger = {
			info: jest.fn(),
			warn: jest.fn(),
			error: jest.fn(),
			debug: jest.fn(),
			isDebugEnabled: jest.fn(() => true),
		} as unknown as winston.Logger;

		mockedPath.__mockPlatform();

		mockedResolver.mockImplementation(createMockResolver(undefined, async () => ({ ec0lint })));

		await new Ec0lintStyleRunner(mockConnection, mockLogger).lintDocument(
			createMockDocument('a {}', '/path/to/file.css'),
			{ config: { rules: { 'block-no-empty': true } } },
		);

		expect(mockLogger.debug).toHaveBeenCalledTimes(1);
		expect(mockLogger.debug).toHaveBeenCalledWith(
			expect.stringMatching(/^Running Ec0lint/),
			expect.any(Object),
		);
	});
});
