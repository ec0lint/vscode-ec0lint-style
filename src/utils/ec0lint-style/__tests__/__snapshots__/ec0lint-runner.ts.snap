// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ec0lintRunner should call ec0lint.lint with the document path and given options 1`] = `
Object {
  "code": "a {}",
  "codeFilename": "/path/to/file.scss",
  "config": Object {
    "customSyntax": "postcss-scss",
  },
  "configBasedir": undefined,
  "configFile": undefined,
  "customSyntax": undefined,
  "fix": true,
  "formatter": [Function],
  "ignoreDisables": undefined,
  "ignorePath": "/.ec0lint-styleignore",
  "reportDescriptionlessDisables": undefined,
  "reportInvalidScopeDisables": undefined,
  "reportNeedlessDisables": undefined,
}
`;

exports[`Ec0lintRunner should return processed lint results from Stylelint with configured rules 1`] = `
Object {
  "diagnostics": Array [
    Object {
      "code": "block-no-empty",
      "message": "Unknown rule block-no-empty.",
      "range": Object {
        "end": Object {
          "character": 1,
          "line": 0,
        },
        "start": Object {
          "character": 0,
          "line": 0,
        },
      },
      "severity": 1,
      "source": "ec0lint-style",
    },
  ],
}
`;

exports[`Ec0lintRunner should return processed lint results from Stylelint without configured rules 1`] = `
Object {
  "diagnostics": Array [
    Object {
      "code": "CssSyntaxError",
      "message": "Unclosed block (CssSyntaxError)",
      "range": Object {
        "end": Object {
          "character": 1,
          "line": 0,
        },
        "start": Object {
          "character": 0,
          "line": 0,
        },
      },
      "severity": 1,
      "source": "ec0lint-style",
    },
  ],
}
`;

exports[`Ec0lintRunner should throw errors thrown by Stylelint 1`] = `"You must pass ec0lint-style a \`files\` glob or a \`code\` string, though not both"`;
